### DISCUSSION TITLE
volume control

### DISCUSSION THREAD

**Discussion by vorosj** [pschatzmann/arduino-audio-tools#1043] [Q&A] [answered]
Hi,
I tried the [streams-i2s-a2dp] example. 

Only mods: uncommented copier.copy(filler) to get both channels,
void loop() {
   copier.copy(filler);
  //copier.copy();
}
and replaced LeftIsEmpty to RightIsEmpty in 
ConverterFillLeftAndRight<int16_t> filler(RightIsEmpty); // fill both channels

It is working, but the volume is low, and the volume control is not working. I get the same sound level whatever I set, for example (0.01 volume is the same as 1)
I set allowBoost=true for values higher then 1, but - as the control is not working at all - of course it doesn't help.

One more question: is it possible to get lower latency? The main source is a buffer, or a codec?

regards, 
   Jozsef

**Comment by pschatzmann** [comment, marked as answer]
In the example the volume.begin() is missing, therefore you don't see any impact!

The stream A2DP examples add an additional layer of buffers with the corresponding lag.
Therefore I am recommending to use the basic API examples instead for A2DP

However don't expect a big difference: A2DP seems to have a big lag because if all the internal bufferung where we do not have any influence. 

ps. I also corrected the bug in the A2DPStream which was setting the initial volume a little too low and removed the VolumeStream from the example to use the built in volume functionality of the A2DP library.  I am aware, that this is not adressing your issue, but it helps to simplify the example. If you don't upgrade, I recommend that you also call a2dpStream,setVolume(1.0) in your sketch.

**Reply by vorosj** [reply]
Thank you for the fast reply.

The base-i2s-a2dp example crashes when connecting. (same hw and same bluetooth headphone is working with the stream example)

_starting A2DP...
Guru Meditation Error: Core  0 panic'ed (LoadProhibited). Exception was unhandled._

platformio.ini:
[env:nodemcu-32s]
platform = espressif32
board = nodemcu-32s
framework = arduino
board_build.partitions = huge_app.csv
monitor_speed = 115200
lib_deps = https://github.com/pschatzmann/arduino-audio-tools
        https://github.com/pschatzmann/ESP32-A2DP.git

**Comment by vorosj** [comment]
I tried the upgradede streams example.
This line has a bug:
 StreamCopy copier(i2sStream, i2sStream); // copy i2sStream to a2dpStream
however, if I replace to a2dpStream, it crashes with the above Guru Meditation Error.

**Comment by pschatzmann** [comment]
It seems that [this example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-basic-api/base-i2s-a2dp/base-i2s-a2dp.ino) got completely messed up. 

I committed the correction that brings it up to what it is supposed to do...

**Comment by vorosj** [comment]
The base-i2s-a2dp is working now, thank you.

Is there a way to add volume control (increasing the volume) to the mic input i2s in the base api? 

Boosting the a2dp output is not the same.
The INMP441 mic has 24 bits resolution. FormatConverterStream conv(i2s) converts the 32bit i2s to 16bit by divide of 2^16, this way it drops 8 useful lsb bits too. It avoids overflow, but loses resolution, and increases noise. Using this mic for human voice, several upper bits are allways zero, but the lost lower bits are valuable.  

I hacked the NumberFormatConverterStreamT class to get higher volume, it's working, but not really elegant solution. :-) 

One more question: how to use filler in the base api? (either on i2s or a2dp, in this case no difference)

**Comment by pschatzmann** [comment]
You can still add the [VolumeStream](https://github.com/pschatzmann/arduino-audio-tools/wiki/Volume-Control) to the chain.
Or you can provide your own conversion using e.g. the more generic CallbackStream.
